strawberryfields/gbs/data.py

```diff
@@ -0,0 +1,215 @@
+:class:`Planted` inherit.
+
+.. autosummary::
+    Dataset
+    Planted
+
+Code details
+^^^^^^^^^^^^
+"""
+# pylint: disable=unnecessary-pass
+from abc import ABC, abstractmethod
+
+import pkg_resources
+import numpy as np
+import scipy
+
+DATA_PATH = pkg_resources.resource_filename("strawberryfields", "gbs/data") + "/"
+
+
+class Dataset(ABC):
+    """Base class for loading datasets of pre-generated samples.
+
+    Attributes:
+        n_mean (float): mean number of photons in the GBS device
+        n_max (float): Maximum number of photons allowed in any sample. This number is set to
+            limit the computation time: any sample being simulated that exceeds ``n_max`` will be
+            ignored and cause the calculation to skip to the next sample.
+        threshold (bool): flag to indicate whether samples are generated with threshold detection
+            (i.e., detectors of zero or some photons) or with photon-number-resolving detectors.
+        adj (array): adjacency matrix of the graph from which samples were generated
+        n_samples (int): total number of samples in the dataset
+        modes (int): number of modes in the GBS device or, equivalently, number of nodes in graph
+        data (sparse): raw data of samples from GBS as a `csr sparse array
+            <https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_matrix.html>`__.
+    """
+
+    _count = 0
+
+    @property
+    @abstractmethod
+    def _data_filename(self) -> str:
+        """Base name of files containing the sample data stored in the ``./data/`` directory.
+
+        Samples should be provided as a ``scipy.sparse.csr_matrix`` saved in ``.npz`` format and
+        the corresponding adjacency matrix should be provided as a ``.npy`` binary.
+
+        For ``_data_filename = "example"``, the corresponding samples should be stored as
+        ``./data/example.npz`` and the adjacency matrix as ``./data/example_A.npy``."""
+        pass
+
+    def __init__(self):
+        self.data = scipy.sparse.load_npz(DATA_PATH + self._data_filename + ".npz")
+        self.adj = np.load(DATA_PATH + self._data_filename + "_A.npy")
+        self.n_samples, self.modes = self.data.shape
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self._count < self.n_samples:
+            self._count += 1
+            return self.__getitem__(self._count - 1)
+        self._count = 0
+        raise StopIteration
+
+    def _elem(self, i):
+        """Access the i-th element of the sparse array and output as a list."""
+        return list(self.data[i].toarray()[0])
+
+    def __getitem__(self, key):
+        if isinstance(key, (slice, tuple)):
+            if isinstance(key, tuple):
+                key = slice(*key)
+            range_tuple = key.indices(self.n_samples)
+            return [self._elem(i) for i in range(*range_tuple)]
+
+        if key < 0:
+            key += self.n_samples
+
+        return self._elem(key)

```

Comment:
Since `__getitem__` can get a variety of inputs, we should probably check the input here and provide some output to the user if they provide an incorrect input (akin to the error you would get from trying to get an element in a list, etc.)

It also could be more readable to flatten the if statements a little, like this:

```python
    def __getitem__(self, key):
        if not isinstance(key, (slice, tuple, int)):
            raise TypeError("Dataset indices must be integers, slices, or tuples")

        if isinstance(key, int):
            return self._elem(key + self.n_samples)
        
        if isinstance(key, tuple):
            key = slice(*key)

        range_tuple = key.indices(self.n_samples)
        return [self._elem(i) for i in range(*range_tuple)]
```

Reply:
![image](https://user-images.githubusercontent.com/49409390/66046846-99808280-e4f4-11e9-8062-96f42c624c6b.png)

^^ That's cool!

Reply:
And thanks for this 😁 Have added!!